mixnet/sim/topology.py

```diff
@@ -0,0 +1,53 @@
+import random
+from collections import defaultdict
+
+from protocol.node import Node
+
+Topology = dict[int, set[int]]
+
+
+def build_full_random_topology(
+    rng: random.Random, num_nodes: int, peering_degree: int
+) -> Topology:
+    """
+    Generate a random undirected topology until all nodes are connected.
+    We don't implement any artificial tool to ensure the connectivity of the topology.
+    Instead, we regenerate a topology in a fully randomized way until all nodes are connected.
+    """
+    while True:
+        topology: Topology = defaultdict(set[int])
+        nodes = list(range(num_nodes))
+        for node in nodes:
+            # Filter nodes that can be connected to the current node.
+            others = []
+            for other in nodes[:node] + nodes[node + 1 :]:
+                if len(topology[other]) < peering_degree:
+                    others.append(other)
+            # How many more connections the current node needs
+            n_needs = peering_degree - len(topology[node])
+            # Sample peers as many as possible
+            peers = rng.sample(others, k=min(n_needs, len(others)))
```

Comment:
We are randomly selecting `peers` here. How do we make sure that the we have not sampled a peer to which we are already connected? Maybe the `.update` call is responsible for removing duplicates?

Reply:
Yes. The `.update` call removes duplicates. However, I just noticed that can make a `node` has less connections than `peering_degree` if there are many duplicates. Let me improve it.

Reply:
Thank you. I fixed it. 61a48b8

With `num_nodes: 100` and `peering_degree: 90` (an extreme case), the difference between the previous algorithm (left) and the fixed algorithm (right) is like below. The fixed algorithm makes node have more connections closer to the `peering_degree: 90`.
![image](https://github.com/user-attachments/assets/68235e92-6a41-418e-aa7f-9a0da99874d3)

In non-extreme cases, such as 100 nodes and 10 degree (or even 50 degree), there was no difference. But, it depends on the seed. The fixed algorithm is better anyway. This fix will affect the current experiment that I'm running (in terms of reproducibility). So, I will not apply this fix to the PR #12 until we have a stable result and analyze it.